apex/transformer/tensor_parallel/layers.py

```diff
@@ -203,15 +223,66 @@ def forward(self, input_):
         return output
 
 
-class ColumnParallelLinearWithAsyncAllreduce(torch.autograd.Function):
-    """
-    Column-parallel linear layer execution with asynchronous all-reduce
-    execution in backprop.
-    """
+class LinearWithGradAccumulationAndAsyncAllreduce(torch.autograd.Function):
+    """Linear layer execution with asynchronous all-reduce and gradient accumulation fusion in backprop."""
+    @staticmethod
+    def forward(ctx, input, weight, bias, gradient_accumulation_fusion, async_grad_allreduce):
+        ctx.save_for_backward(input, weight)
+        ctx.use_bias = bias is not None
+        ctx.gradient_accumulation_fusion = gradient_accumulation_fusion
+        ctx.async_grad_allreduce = async_grad_allreduce
+        output = torch.matmul(input, weight.t())
+        if bias is not None:
+            output = output + bias
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input, weight = ctx.saved_tensors
+        use_bias = ctx.use_bias
+        grad_input = grad_output.matmul(weight)
+        # Convert the tensor shapes to 2D for execution compatibility
+        grad_output = grad_output.view(grad_output.shape[0] * grad_output.shape[1], grad_output.shape[2])
+        input = input.view(input.shape[0] * input.shape[1], input.shape[2])
+        if ctx.async_grad_allreduce:
+            # Asynchronous all-reduce
+            handle = torch.distributed.all_reduce(grad_input, group=get_tensor_model_parallel_group(), async_op=True)
+            # Delay the start of weight gradient computation shortly (3us) to have
+            # all-reduce scheduled first and have GPU resources allocated
+            _ = torch.empty(1, device=grad_output.device) + 1
```

Comment:
Is this guaranteed to cause the `all_reduce` to happen first?

Reply:
experimentally yes.